### PR TITLE
Feature/#10100 UI list of readers

### PR DESCRIPTION
**Pull request by ovidiuvladpop**

### CODE REVIEW

File: ClearentIdtechIOSFramework/ClearentUI/ClearentProcessingModalViewController.swift [hunk 1]
```diff
@@ -45,37 +62,80 @@ extension ClearentProcessingModalViewController: ClearentProcessingModalView {
...
+        if let _ = presenter?.getSelectedReaderFromReadersList() {
+            signalStatus = readerInfo.signalStatus(flowFeedbackType: flowFeedbackType, isConnecting: true)
+        } else {
+            signalStatus = readerInfo.signalStatus(flowFeedbackType: flowFeedbackType)
+        }
         let batteryStatus = readerInfo.batteryStatus(flowFeedbackType: flowFeedbackType)
         let statusHeader = ClearentReaderStatusHeaderView()
-        statusHeader.setup(readerName: name, signalStatusIconName: signalStatus.iconName, signalStatusTitle: signalStatus.title, batteryStatusIconName: batteryStatus.iconName, batteryStatusTitle: batteryStatus.title)
+        
+        showOnTop ? statusHeader.setup(readerName: name, dropDownIconName: ClearentConstants.IconName.expanded, signalStatusIconName: signalStatus.iconName, signalStatusTitle: signalStatus.title, batteryStatusIconName: batteryStatus.iconName, batteryStatusTitle: batteryStatus.title) : statusHeader.setup(readerName: name, signalStatusIconName: signalStatus.iconName, signalStatusTitle: signalStatus.title, batteryStatusIconName: batteryStatus.iconName, batteryStatusTitle: batteryStatus.title)
```
Comment (carmenj) [inline ClearentIdtechIOSFramework/ClearentUI/ClearentProcessingModalViewController.swift:132]: This line is hard to read. Could be simplified like this.
```suggestion
let dropDownIconName = showOnTop ? ClearentConstants.IconName.expanded : nil
statusHeader.setup(readerName: name, dropDownIconName: dropDownIconName, signalStatusIconName: signalStatus.iconName, signalStatusTitle: signalStatus.title, batteryStatusIconName: batteryStatus.iconName, batteryStatusTitle: batteryStatus.title)
```
You can also set fewer parameters to the `setup` method: only `readerName`, `dropDownIconName`, `signalStatus` & `batteryStatus`

File: ClearentIdtechIOSFramework/ClearentUI/Common/Assets/Localizable.strings [hunk 2]
```diff
@@ -13,7 +13,7 @@
 "xsdk_press_reader_button" = "Press the button on reader to reconnect";
 "xsdk_select_reader" = "Please select the reader you wish to pair with";
 "xsdk_connecting_reader" = "Connecting...";
-"xsdk_connection_sucesfull" = "Connection successful";
+"xsdk_connection_sucessful" = "Connection successful";
```
Comment (carmenj) [inline ClearentIdtechIOSFramework/ClearentUI/Common/Assets/Localizable.strings:16]: successful 😀

File: ClearentIdtechIOSFramework/ClearentUI/Common/UIComponents/ClearentReaderStatusHeaderView.swift [hunk 3]
```diff
@@ -25,17 +30,37 @@ public class ClearentReaderStatusHeaderView: ClearentMarginableView {
...
             readerBatteryStatusView.isHidden = false
             readerBatteryStatusView.setup(imageName: batteryImage, status: batteryStatus)
             return
         }
         readerBatteryStatusView.isHidden = true
     }
+    
+    public func updateDropDownIcon() {
+        dropDownImageView.image = state == .collapsed ? UIImage(named: ClearentConstants.IconName.collapsed, in: ClearentConstants.bundle, compatibleWith: nil) :
+        UIImage(named: ClearentConstants.IconName.expanded, in: ClearentConstants.bundle, compatibleWith: nil)
```
Comment (carmenj) [inline ClearentIdtechIOSFramework/ClearentUI/Common/UIComponents/ClearentReaderStatusHeaderView.swift:55]:
```suggestion
        let iconName = state == .collapsed ? ClearentConstants.IconName.collapsed : ClearentConstants.IconName.expanded
        dropDownImageView.image =  UIImage(named: iconName, in: ClearentConstants.bundle, compatibleWith: nil)
```

File: ClearentIdtechIOSFramework/ClearentUI/PaymentFlowDataSource/FlowDataProvider.swift [hunk 4]
```diff
@@ -210,15 +211,35 @@ extension FlowDataProvider : ClearentWrapperProtocol {
...
+        let feedback: FlowFeedback
+        
+        if ClearentWrapper.shared.flowType == .showReaders {
+            guard var recentlyPairedReaders = ClearentWrapperDefaults.recentlyPairedReaders else { return }
+            guard let pairedReaderInfo = ClearentWrapperDefaults.pairedReaderInfo else { return }
+            guard let indexOfSelectedReader = recentlyPairedReaders.firstIndex(where: {$0.readerName == pairedReaderInfo.readerName}) else { return }
+            
+            recentlyPairedReaders[indexOfSelectedReader].signalLevel = pairedReaderInfo.signalLevel
+            recentlyPairedReaders[indexOfSelectedReader].batterylevel = pairedReaderInfo.batterylevel
+            recentlyPairedReaders[indexOfSelectedReader].isConnected = pairedReaderInfo.isConnected
```
Comment (carmenj) [inline ClearentIdtechIOSFramework/ClearentUI/PaymentFlowDataSource/FlowDataProvider.swift:224]: Can these lines be replaced with  `recentlyPairedReaders[indexOfSelectedReader] = pairedReaderInfo`?

File: ClearentIdtechIOSFramework/ClearentUI/ClearentProcessingModalViewController.swift [hunk 5]
```diff
@@ -45,37 +62,80 @@ extension ClearentProcessingModalViewController: ClearentProcessingModalView {
...
         case .hint:
             guard let text = object as? String else { return nil }
+            
             return ClearentHintView(text: text)
+        case .recentlyPaired:
+            guard let readersInfo = object as? [ReaderInfo] else { return nil }
+            guard let pairedReaderInfo = ClearentWrapperDefaults.pairedReaderInfo else { return nil }
+            var readersTableViewDataSource: [ReaderItem] = readersInfo.map { item -> ReaderItem in
+                    return ReaderItem(readerInfo: item)
+            }
```
Comment (carmenj) [inline ClearentIdtechIOSFramework/ClearentUI/ClearentProcessingModalViewController.swift:100]:
```suggestion
    var readersTableViewDataSource: [ReaderItem] = readersInfo.map { ReaderItem(readerInfo: $0) }
```

**Review by carmenj — CHANGES_REQUESTED** [review]

File: ClearentIdtechIOSFramework/ClearentUI/ClearentProcessingModalPresenter.swift [hunk 6]
```diff
@@ -17,6 +17,10 @@ public protocol ClearentProcessingModalView: AnyObject {
 public protocol ProcessingModalProtocol {
     func restartProcess(processType: ProcessType)
     func startFlow()
+    func startPairingFlow()
+    func connectTo(reader: ReaderInfo)
+    func getSelectedReaderFromReadersList() -> ReaderItem?
+    func setSelectedReaderFromReadersList(_ readerItem: ReaderItem?)
```
Comment (carmenj) [inline ClearentIdtechIOSFramework/ClearentUI/ClearentProcessingModalPresenter.swift:23]: couldn't we replace these 2 methods with a single computed property that has a setter and getter?

Comment (ovidiuvladpop) [reply ClearentIdtechIOSFramework/ClearentUI/ClearentProcessingModalPresenter.swift:23]: Updated

File: ClearentIdtechIOSFramework/ClearentUI/ClearentProcessingModalViewController.swift (hunk 5 above)
Comment (ovidiuvladpop) [reply ClearentIdtechIOSFramework/ClearentUI/ClearentProcessingModalViewController.swift:100]: Updated

File: ClearentIdtechIOSFramework/ClearentUI/PaymentFlowDataSource/FlowDataProvider.swift (hunk 4 above)
Comment (ovidiuvladpop) [reply ClearentIdtechIOSFramework/ClearentUI/PaymentFlowDataSource/FlowDataProvider.swift:224]: Updated

**Review by carmenj — APPROVED** [review]

File: ClearentIdtechIOSFramework/ClearentUI/ClearentProcessingModalViewController.swift [hunk 7]
```diff
@@ -45,37 +62,75 @@ extension ClearentProcessingModalViewController: ClearentProcessingModalView {
...
+                
+                return ClearentReadersTableView(dataSource: readersTableViewDataSource, delegate: self)
+            }
         }
     }
 
     private func readerInfoView(readerInfo: ReaderInfo, flowFeedbackType: FlowFeedbackType) -> ClearentReaderStatusHeaderView {
         let name = readerInfo.readerName
-        let signalStatus = readerInfo.signalStatus(flowFeedbackType: flowFeedbackType)
+        let signalStatus = presenter?.selectedReaderFromReadersList != nil ? readerInfo.signalStatus(flowFeedbackType: flowFeedbackType, isConnecting: true) : readerInfo.signalStatus(flowFeedbackType: flowFeedbackType)
```
Comment (carmenj) [inline ClearentIdtechIOSFramework/ClearentUI/ClearentProcessingModalViewController.swift:122]:
```suggestion
        let signalStatus =  readerInfo.signalStatus(flowFeedbackType: flowFeedbackType, isConnecting: presenter?.selectedReaderFromReadersList != nil)
```

File: ClearentIdtechIOSFramework/ClearentUI/ClearentProcessingModalPresenter.swift [hunk 8]
```diff
@@ -25,6 +28,7 @@ public class ClearentProcessingModalPresenter {
     private let sdkWrapper = ClearentWrapper.shared
     private var sdkFeedbackProvider: FlowDataProvider
     private let processType: ProcessType
+    public var selectedReaderFromReadersList: ReaderItem?
```
Comment (carmenj) [inline ClearentIdtechIOSFramework/ClearentUI/ClearentProcessingModalPresenter.swift:31]: should `selectedReaderFromReadersList` be `public`? maybe `internal` is good enough

**Pull request merged**